### PR TITLE
[ready] Turn Lane Guidance

### PR DESCRIPTION
**Pull request by MoKob**
Implementing https://github.com/Project-OSRM/osrm-backend/issues/2296

- [x] extraction
    - [x] parse turn lane data during extraction
          - [x] turn:lane*
          - [x] destination:lanes*
    - [x] match turn lane string to possible turns
    - [x] create turn lane tupels (lanes at turn, lanes possible for instruction, first possible lane from the right)
    - [x] get turn lanes through data pipeline

- [x] routing
    - [x] expose available lanes to the API
    - [x] implement concept of independent turns
    - [x] implement post-processing to restrict multiple possible lanes to fewer lanes, when needed

- [x] cucumber
    - [x] expose lane guidance
    - [x] write lane guidance tests

- [x] docs/changelog
    - [x] document changes in doc/http.md
    - [x] update changelog

### CODE REVIEW

File: include/extractor/guidance/toolkit.hpp [hunk 1]
```diff
@@ -470,6 +470,52 @@ inline ConnectedRoad mirror(ConnectedRoad road)
...
+        for (std::size_t i = 0; i < count_left; ++i)
+            // this is adjusted for our fake pipe. The moment cucumber can handle multiple escaped
+            // pipes, the '&' part can be removed
+            if (lane_string[i] != '|' && lane_string[i] != '&')
+            {
+                sane = false;
+                break;
+            }
+
+        if( sane )
```
Comment (daniel-j-h) [inline include/extractor/guidance/toolkit.hpp]: clang-format38

File: include/util/guidance/turn_lanes.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+} // namespace std
+
+namespace osrm
+{
+namespace util
+{
+namespace guidance
+{
+
+// The mapping of turn lanes can be done using a tupel as described here.
```
Comment (daniel-j-h) [inline include/util/guidance/turn_lanes.hpp]: Keeping track of this so we don't forget about it before merging:

- [ ] remove outdated example or adapt to new representation

File: src/contractor/contractor.cpp [hunk 3]
```diff
@@ -76,7 +76,7 @@ int Contractor::Run()
 #ifdef WIN32
 #pragma message("Memory consumption on Windows can be higher due to different bit packing")
 #else
-    static_assert(sizeof(extractor::NodeBasedEdge) == 20,
+    static_assert(sizeof(extractor::NodeBasedEdge) == 24,
                   "changing extractor::NodeBasedEdge type has influence on memory consumption!");
```
Comment (MoKob) [inline src/contractor/contractor.cpp]: windows size might be off

File: include/extractor/guidance/turn_instruction.hpp [hunk 4]
```diff
@@ -146,16 +147,18 @@ struct TurnInstruction
     }
 };
 
-static_assert(sizeof(TurnInstruction) == 1, "TurnInstruction does not fit one byte");
+static_assert(sizeof(TurnInstruction) == 3, "TurnInstruction does not fit three byte");
```
Comment (MoKob) [inline include/extractor/guidance/turn_instruction.hpp]: its currently a typo, but @daniel-j-h might have to modify this anyways. @daniel-j-h can you check this size when you have the switch and make sure to fix the typo when you are on it?

File: .clang-format [hunk 5]
```diff
@@ -2,7 +2,7 @@
 Language:        Cpp
 # BasedOnStyle:  LLVM
 AccessModifierOffset: -2
-AlignAfterOpenBracket: Align
+AlignAfterOpenBracket: true
```
Comment (TheMarex) [inline .clang-format]: Where are these changes coming from? This should be at least an own commit.

File: docs/http.md [hunk 6]
```diff
@@ -437,30 +437,42 @@ step.
...
+- `lanes`: the available turn lanes at the turn
+  - `marked`: markings on the road, following the OSM scheme (e.g. left, slight_right, or through)
+  - `take`: a boolean flag indicating whether the lane is a possible choice in the maneuver
 - `pronunciation`: The pronunciation hint of the way name. Will be `undefined` if there is no pronunciation hit.
 - `destinations`: The destinations of the way. Will be `undefined` if there are no destinations.
 - `mode`: A string signifying the mode of transportation.
 - `maneuver`: A `StepManeuver` object representing the maneuver.
 - `intersections`: A list of `Intersections` that are passed along the segment, the very first belonging to the StepManeuver
 
+Currently, the supported lane tags are: `sharp_left, left, slight_left, sharp_right, right, slight_right, through, reverse, none`.
```
Comment (TheMarex) [inline docs/http.md]: This needs to be a table like the turn type with descriptions for each.

File: docs/http.md [hunk 7]
```diff
@@ -437,30 +437,42 @@ step.
   | geojson    | [GeoJSON `LineString`](http://geojson.org/geojson-spec.html#linestring) or [GeoJSON `Point`](http://geojson.org/geojson-spec.html#point) if it is only one coordinate (not wrapped by a GeoJSON feature)|
   
 - `name`: The name of the way along which travel proceeds.
+- `lanes`: the available turn lanes at the turn
+  - `marked`: markings on the road, following the OSM scheme (e.g. left, slight_right, or through)
```
Comment (TheMarex) [inline docs/http.md]: `marked` is a strange property name for this. If its marking on the road or a sign does not really matter, the interesting think is what type of lane this is. As such I would propose `type`.

File: docs/http.md [hunk 8]
```diff
@@ -437,30 +437,42 @@ step.
   | geojson    | [GeoJSON `LineString`](http://geojson.org/geojson-spec.html#linestring) or [GeoJSON `Point`](http://geojson.org/geojson-spec.html#point) if it is only one coordinate (not wrapped by a GeoJSON feature)|
   
 - `name`: The name of the way along which travel proceeds.
+- `lanes`: the available turn lanes at the turn
```
Comment (TheMarex) [inline docs/http.md]: Are these only valid at the turn point or for the whole segment? If it is only valid at the turn this need to be part of `StepManeuver`.

File: docs/http.md [hunk 9]
```diff
@@ -437,30 +437,42 @@ step.
...
+  - `marked`: markings on the road, following the OSM scheme (e.g. left, slight_right, or through)
+  - `take`: a boolean flag indicating whether the lane is a possible choice in the maneuver
```
Comment (TheMarex) [inline docs/http.md]: Should be named `valid`.

File: docs/http.md [hunk 10]
```diff
@@ -437,30 +437,42 @@ step.
...
+Currently, the supported lane tags are: `sharp_left, left, slight_left, sharp_right, right, slight_right, through, reverse, none`.
+
 #### Example
 
 ```
 {
  "distance":152.3,
  "duration":15.6,
  "name":"Lortzingstraße",
+ "lanes":{
```
Comment (TheMarex) [inline docs/http.md]: `lanes` is and Array and this needs to use `[]` braces.

File: include/extractor/guidance/debug.hpp [hunk 11]
```diff
@@ -0,0 +1,46 @@
+#ifndef OSRM_EXTRACTOR_GUIDANCE_DEBUG_HPP_
```
Comment (TheMarex) [inline include/extractor/guidance/debug.hpp]: Can we put all this debug logic in a central debug header (`util/debug.hpp`)? We don't need to be worried about a star topology with dependencies since this header should never be included in production code.

**Comment by TheMarex** [comment]
I started a review on this PR. I already added some smaller suggestions. Will post back tomorrow with some more involved questions/concerns.

File: docs/http.md (hunk 6 above)
Comment (MoKob) [inline docs/http.md]: What kind of description would you expect? This would be: left - an arrow pointing to the left. Would a link to the OSM definition be acceptable? These are directly passed through from OSM.

File: .clang-format (hunk 5 above)
Comment (MoKob) [inline .clang-format]: This fixes an error about an unsupported value for this choice. Without it clang-format reformats close to the entire code-base for me. Without, I only get a format of my own files.

Will open a dedicated PR for it.

Comment (MoKob) [inline .clang-format]: Hm. I was getting all kinds with YAML errors with this. No idea why I can't reproduce it right now. Removing

File: docs/http.md (hunk 7 above)
Comment (MoKob) [inline docs/http.md]: Well. Its really just the marking on the road. So `lane` does not really fit in my view. Would `marking` be more to your liking?

File: docs/http.md (hunk 8 above)
Comment (MoKob) [inline docs/http.md]: This is in fact a documentation mix-up. They are part of `StepManeuver` in the code already.

**Comment by TheMarex** [comment]
Capturing from chat:

- [ ] Loading the original lane tag information as strings into the server needs to go
- [ ] The lane tags will be replaced by normalized enums and not just pass through OSM values
- [ ] Instead of using a `std::vector<char>` for storing lane tags this means it will be `std::vector<LaneType>`

File: docs/http.md [hunk 12]
```diff
@@ -481,6 +487,13 @@ step.
   direction of travel immediately after the maneuver.
 - `type` A string indicating the type of maneuver. **new identifiers might be introduced without API change**
    Types  unknown to the client should be handled like the `turn` type, the existance of correct `modifier` values is guranteed.
+- `lanes`: the available turn lanes at the turn
+  - `indication`: markings on the road, following the OSM scheme (e.g. left, slight_right, or through). Markings mey be combined into a list separated by ';'.
+  - `valid`: a boolean flag indicating whether the lane is a valid choice in the maneuver
+
+Currently, the supported lane tags are: `sharp_left, left, slight_left, sharp_right, right, slight_right, through, reverse, none`.
```
Comment (TheMarex) [inline docs/http.md]: We should make this consistent with our modifiers. We don't use underscore there. Suggestion:

| `direction`      | Description                               |
| ------------------- | ---------------------------------------|
| uturn             | lane used for uturn maneuver          |
| sharp right       | lane for a sharp right turn                        |
| right             | lane for  a normal turn to the right                |
| slight right      | lane for a slight turn to the right                |
| straight          | lane to go straight           |
| slight left       | lane for a slight turn to the left                 |
| left              | lane for a normal turn to the left                 |
| sharp left        | lane for a sharp turn to the left                  |
| all | lane allows turns in all direction |

We should also change `indication` -> `direction` or `type`.

**Comment by TheMarex** [comment]
API response for a `RouteStep` will look like this if my current suggestions are implemented:

```
{
 "distance":152.3,
 "duration":15.6,
 "name":"Lortzingstraße",
 "maneuver":{
     "type":"turn",
     "modifier":"right",
     "lanes":[
         {"direction":"sharp left", "valid":"false"},
         {"direction":"straight", "valid":"true"},
         {"direction":"right", "valid":"true"}
     ]},
 "geometry":"{lu_IypwpAVrAvAdI",
 "mode":"driving",
 "intersections":[
    {"location":[13.39677,52.54366],
    "in":2,
    "out":1,
    "bearings":[10,184,270],
    "entry":[false,"true","true"]},
    {"location":[13.394718,52.543096],
    "in":0,
    "out":2,
    "bearings":[60,150,240,330],
    "entry":["false","true","true","true"]
    }
]}
```

File: docs/http.md (hunk 12 above)
Comment (MoKob) [inline docs/http.md]: In my current rewrite I already have the variation without underscore.
I feel that `all` is not a valid representation, because even if there is no marking (which would be all), it may only be allowed to go straight. The markings here represent the actual markings on the street.

The same way I would feel badly about direction. What we report here is nothing more than an indication, which is the current name. Direction would mean that it can be used only for a specific direction. A slight-right arrow can be used for exiting a round-about in two ways, for example.
We could go for type, but it would be confusing, since it is not a turn-road. So type would feel to me that it is a dedicated (separated) road. But as I said, indication is all we get.

File: docs/http.md [hunk 13]
```diff
@@ -481,6 +485,10 @@ step.
   direction of travel immediately after the maneuver.
 - `type` A string indicating the type of maneuver. **new identifiers might be introduced without API change**
    Types  unknown to the client should be handled like the `turn` type, the existance of correct `modifier` values is guranteed.
+- `lanes`: the available turn lanes at the turn
+  - `indication`: markings on the road, following the OSM scheme (e.g. left, slight_right, or through). Markings mey be combined into a list separated by ';'.
```
Comment (TheMarex) [inline docs/http.md]: > Markings mey be combined into a list separated by ';'.

I don't think this documentation is valid anymore. The Table needs to be moved up here.

File: docs/http.md [hunk 14]
```diff
@@ -535,6 +543,20 @@ step.
   | else                   | Indicates the number of intersections passed until the turn. Example instruction: `at the fourth intersection, turn left` |
   
 
+- `indication`
+  
+  | `type`                 | Description                                                                                                               |
```
Comment (TheMarex) [inline docs/http.md]: I think this table is broken. Member should be named `indication` to be consistent with your other namings.

File: include/engine/datafacade/shared_datafacade.hpp [hunk 15]
```diff
@@ -79,18 +81,25 @@ class SharedDataFacade final : public BaseDataFacade
...
     util::ShM<NodeID, true>::vector m_via_node_list;
     util::ShM<unsigned, true>::vector m_name_ID_list;
+    util::ShM<LaneDataID, true>::vector m_lane_data_id;
     util::ShM<extractor::guidance::TurnInstruction, true>::vector m_turn_instruction_list;
     util::ShM<extractor::TravelMode, true>::vector m_travel_mode_list;
     util::ShM<char, true>::vector m_names_char_list;
+    util::ShM<char, true>::vector m_turn_string_char_list;
     util::ShM<unsigned, true>::vector m_name_begin_indices;
+    util::ShM<char, true>::vector m_lane_string_char_list;
+    util::ShM<unsigned, true>::vector m_lane_string_begin_indices;
```
Comment (TheMarex) [inline include/engine/datafacade/shared_datafacade.hpp]: These are left-over from the old code.

File: include/engine/datafacade/shared_datafacade.hpp [hunk 16]
```diff
@@ -79,18 +81,25 @@ class SharedDataFacade final : public BaseDataFacade
     util::PackedVector<OSMNodeID, true> m_osmnodeid_list;
     util::ShM<NodeID, true>::vector m_via_node_list;
     util::ShM<unsigned, true>::vector m_name_ID_list;
+    util::ShM<LaneDataID, true>::vector m_lane_data_id;
     util::ShM<extractor::guidance::TurnInstruction, true>::vector m_turn_instruction_list;
     util::ShM<extractor::TravelMode, true>::vector m_travel_mode_list;
     util::ShM<char, true>::vector m_names_char_list;
+    util::ShM<char, true>::vector m_turn_string_char_list;
```
Comment (TheMarex) [inline include/engine/datafacade/shared_datafacade.hpp]: Also not used anymore.

File: include/engine/datafacade/internal_datafacade.hpp [hunk 17]
```diff
@@ -78,14 +79,19 @@ class InternalDataFacade final : public BaseDataFacade
     util::ShM<NodeID, false>::vector m_via_node_list;
     util::ShM<unsigned, false>::vector m_name_ID_list;
     util::ShM<extractor::guidance::TurnInstruction, false>::vector m_turn_instruction_list;
+    util::ShM<LaneDataID, false>::vector m_lane_data_id;
+    util::ShM<util::guidance::LaneTupelIdPair, false>::vector m_lane_tupel_id_pairs;
     util::ShM<extractor::TravelMode, false>::vector m_travel_mode_list;
     util::ShM<char, false>::vector m_names_char_list;
+    util::ShM<char, false>::vector m_lanes_char_list;
```
Comment (TheMarex) [inline include/engine/datafacade/internal_datafacade.hpp]: Not used anymore.

File: include/extractor/guidance/turn_lane_types.hpp [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+namespace TurnLaneType
+{
+namespace detail
+{
+const constexpr std::size_t num_supported_lane_types = 11;
+} // namespace detail
+
+typedef std::uint16_t Mask;
+const constexpr Mask empty = 0;
+const constexpr Mask none = 1 << 0;
```
Comment (TheMarex) [inline include/extractor/guidance/turn_lane_types.hpp]: When using bitshift operators it is good to use the unsigned post-fixes for numbers to make sure we don't accidentally introduce undefined behavior by shifting into the sign bit. Like: `1u << 0u`

File: include/extractor/guidance/toolkit.hpp [hunk 19]
```diff
@@ -476,6 +483,77 @@ inline bool hasRoundaboutType(const TurnInstruction instruction)
...
+// are not specifically marked for left/right turns. This function can be used from the profile to
+// trim the lane string appropriately
+//
+// left|throught|
+// in combination with lanes:psv:forward=1
+// will be corrected to left|throught, since the final lane is not drivable.
+// This is in contrast to a situation with lanes:psv:forward=0 (or not set) where left|through|
+// represents left|through|through
+inline std::string
+trimLaneString(std::string lane_string, std::int32_t count_left, std::int32_t count_right)
```
Comment (TheMarex) [inline include/extractor/guidance/toolkit.hpp]: Is this function unit tested? This seems like a good candidate for some hardening because it handles user input.

File: include/extractor/guidance/turn_lane_matcher.hpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+// the best possible match for every tag in a given intersection, considering a few corner cases
+// introduced to OSRM handling u-turns
+typename Intersection::const_iterator findBestMatch(const TurnLaneType::Mask &tag,
+                                                    const Intersection &intersection);
+// Reverse is a special case, because it requires access to the leftmost tag. It has its own
+// matching function as a result of that. The leftmost tag is required, since u-turns are disabled
+// by default in OSRM. Therefor we cannot check whether a turn is allowed, since it could be
+// possible that it is forbidden. In addition, the best u-turn angle does not necessarily represent
+// the u-turn, since it could be a sharp-left turn instead on a road with a middle island.
+typename Intersection::const_iterator findBestMatchForReverse(const TurnLaneType::Mask &leftmost_tag,
```
Comment (TheMarex) [inline include/extractor/guidance/turn_lane_matcher.hpp]: Minor nit-pick I would put these (good!) documentation where the code lives not in the definition header.

File: include/storage/shared_datatype.hpp [hunk 21]
```diff
@@ -113,11 +151,13 @@ struct SharedDataLayout
             bool end_canary_alive = std::equal(CANARY, CANARY + sizeof(CANARY), end_canary_ptr);
             if (!start_canary_alive)
             {
-                throw util::exception("Start canary of block corrupted.");
+                throw util::exception(std::string("Start canary of block corrupted. (") +
```
Comment (TheMarex) [inline include/storage/shared_datatype.hpp]: Hah good change. :+1:

File: include/storage/shared_datatype.hpp [hunk 22]
```diff
@@ -49,6 +83,10 @@ struct SharedDataLayout
         BEARING_BLOCKS,
         BEARING_VALUES,
         ENTRY_CLASS,
+        LANE_DATA_ID,
+        TURN_LANE_DATA,
+        LANE_DESCRIPTION_OFFSETS,
+        LANE_DESCRIPTION_MASKS,
```
Comment (TheMarex) [inline include/storage/shared_datatype.hpp]: Add an static_assert here that check that `NUM_BLOCKS` and the size of `block_id_to_name` are the same. otherwise we might have errors in the error reporting.

File: include/util/debug.hpp [hunk 23]
```diff
@@ -0,0 +1,88 @@
+#ifndef OSRM_ENGINE_GUIDANCE_DEBUG_HPP_
+#define OSRM_ENGINE_GUIDANCE_DEBUG_HPP_
+
```
Comment (TheMarex) [inline include/util/debug.hpp]: Jup this is a good change. We have lots of other debug code (all the operator<< implementations for various types) that we could move in here.

File: include/util/group_by.hpp [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+{
+
+// Runs fn on consecutive items in sub-ranges determined by pred.
+//
+// Example:
+//   vector<int> v{1,2,2,2,3,4,4};
+//   group_by(first, last, even, print);
+//   >>> 2,2,2
+//   >>> 4,4
+template <typename Iter, typename Pred, typename Fn>
```
Comment (TheMarex) [inline include/util/group_by.hpp]: Is there a unit-test for this?

File: src/engine/api/json_factory.cpp [hunk 25]
```diff
@@ -56,17 +57,43 @@ const constexpr char *waypoint_type_names[] = {"invalid", "arrive", "depart"};
...
     return turn_type_names[static_cast<std::size_t>(type)];
 }
 
+util::json::Array lanesFromManeuver(const guidance::StepManeuver &maneuver)
+{
+    BOOST_ASSERT(maneuver.lanes.lanes_in_turn >= 1);
+    util::json::Array result;
+    LaneID lane_id = maneuver.lane_description.size();
+
+    for (auto iter = maneuver.lane_description.begin(); iter != maneuver.lane_description.end(); ++iter)
```
Comment (TheMarex) [inline src/engine/api/json_factory.cpp]: This can be a for-each loop now: `for(auto desc : maneuver.lane_description)`

File: src/engine/guidance/post_processing.cpp [hunk 26]
```diff
@@ -31,16 +32,18 @@ namespace guidance
 
 namespace
 {
-const constexpr double MAX_COLLAPSE_DISTANCE = 25;
+const constexpr double MAX_COLLAPSE_DISTANCE = 30;
```
Comment (TheMarex) [inline src/engine/guidance/post_processing.cpp]: Why the constant changes here?

File: src/engine/guidance/post_processing.cpp [hunk 27]
```diff
@@ -406,6 +408,7 @@ void collapseTurnAt(std::vector<RouteStep> &steps,
     else if (one_back_step.distance <= MAX_COLLAPSE_DISTANCE &&
              isCollapsableInstruction(current_step.maneuver.instruction))
     {
+        // TODO check for lanes
```
Comment (TheMarex) [inline src/engine/guidance/post_processing.cpp]: What is up with this TODO?

File: include/util/group_by.hpp (hunk 24 above)
Comment (daniel-j-h) [inline include/util/group_by.hpp]: For the record, this is basically
https://docs.python.org/3.5/library/itertools.html#itertools.groupby

File: src/engine/api/json_factory.cpp (hunk 25 above)
Comment (daniel-j-h) [inline src/engine/api/json_factory.cpp]: => `scripts/modernize.sh`

**Comment by TheMarex** [comment]
@MoKob looks much better. :+1: Only remaining issues are outdated docs and some minor stuff.

File: include/util/group_by.hpp (hunk 24 above)
Comment (TheMarex) [inline include/util/group_by.hpp]: Yeah we can add the tests in another PR, this is not blocking. But it would make sense to have tests for utility functions.

File: src/engine/guidance/post_processing.cpp (hunk 27 above)
Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: This is related to https://github.com/Project-OSRM/osrm-backend/issues/2553

File: src/engine/guidance/post_processing.cpp (hunk 26 above)
Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: Well. Thats the thing with arbitrary constants. Some time you find out that your first guess wasn't large enough.

File: include/extractor/guidance/toolkit.hpp (hunk 19 above)
Comment (MoKob) [inline include/extractor/guidance/toolkit.hpp]: since recently (@danpat and my late night debugging session) it has a dedicated cucumber test for its edge-cases

File: include/storage/shared_datatype.hpp (hunk 22 above)
Comment (MoKob) [inline include/storage/shared_datatype.hpp]: done.

File: docs/http.md [hunk 28]
```diff
@@ -481,6 +485,10 @@ step.
   direction of travel immediately after the maneuver.
 - `type` A string indicating the type of maneuver. **new identifiers might be introduced without API change**
    Types  unknown to the client should be handled like the `turn` type, the existance of correct `modifier` values is guranteed.
+- `lanes`: the available turn lanes at the turn
+  - `indication`: a indication (e.g. marking on the road) specifying the turn lane. Multiple indications may be combined with `;` (e.g. straight;left).
```
Comment (TheMarex) [inline docs/http.md]: @MoKob is it still the case that ```multiple indications may be combined with `;````?

**Comment by TheMarex** [comment]
There is just the one clarification I would like to have with regard to multiple indications. Otherwise we can merge this as soon as possible.

If there can be multiple indications per lane (why?), then we would need to change the API response format because of the same anti-pattern as before: Arrays encoded as strings in JSON.

So it would be:
```json
{
   "indications": ["left", "straight"],
   "valid": true,
}
```

**Comment by MoKob** [comment]
The why is combination markers: arrow for going straight and going left.

**Comment by TheMarex** [comment]
@mokob looks good now. We can merge once travis is green.
